### ISSUE TITLE
Avoid the limit of ROI annotation of 5000x5000 px²

### ISSUE THREAD

**Issue by geektortoise** [cytomine/S_CellDetect_Stardist_HE_ROI#2]
With huge ROI annotation, we can have
- timeout from IMS
- error from IMS
or
- error from tensorflow

One way to improve the script is to detect if it is a large annotation and split the ROI zone into multiples ones ?

**Comment by mizjaggy18** [comment]
Hi @geektortoise, 

I'm facing the same issue now. No timeout error but the annotation didn't scale correctly. I created a large ROI, as in the photo below. It can run the algorithm successfully but the detection area was scaled down. The annotations also did not overlap on any of the nuclei but its overall shape matches the large ROI. 

![image](https://user-images.githubusercontent.com/77017051/120107375-6bed5000-c193-11eb-8780-115f210fc85e.png)
![image](https://user-images.githubusercontent.com/77017051/120107582-1e251780-c194-11eb-9876-81318f36507a.png)


When selecting smaller ROI, the algorithm can do the detection very good: 
![image](https://user-images.githubusercontent.com/77017051/120107515-e0c08a00-c193-11eb-8b40-8e05a348a4ed.png)
![image](https://user-images.githubusercontent.com/77017051/120107605-46ad1180-c194-11eb-9325-7930dc81b6d7.png)

Is there any way to automatically divide the large ROI intro several smaller ROIs?

Thank you.

**Comment by anttikos** [comment]
Hi.

I faced the same issue that was reported by @mizjaggy18 in the previous message, when I tested with the latest version of Cytomine (at the time of writing this). After a bit of digging, I noticed that the cropped image, which is used for the actual cell detection, has max size limit of 10000x10000 pixels. If the original ROI is bigger than this, the cell annotations are not scaled up correctly in relation to the original ROI size. In short, changing the line 121 of run.py
 
from:
`p = Point(minx+polygroup[1][i],miny-polygroup[0][i])`

to:
`p = Point(minx+(polygroup[1][i]*((roi_geometry.bounds[2]-minx)/im.size[0])),miny-(polygroup[0][i]*((miny-roi_geometry.bounds[1])/im.size[1])))`

...seems to fix the issue, and works well at least on my use case/data set. Other possible fix would be to remove the 10000x10000 pixel limit from the crop function (to avoid any possible data loss/inaccuracy caused by resizing the cropped area), although this would slow down the crop function, and might required some tweaking on other parts of Cytomine as well.

Also, there was a timeout issue when the app was calling the crop API, but that got solved by changing the Cytomine configs and setting the nginx http timeout to:
`    proxy_read_timeout 600;
    proxy_connect_timeout 600;
    proxy_send_timeout 600;
`
